### ISSUE TITLE
Bug: [no-unnecessary-condition] Does not account for optional chaining on potentially-`void` values

### ISSUE THREAD

**Issue by RedGuy12** [typescript-eslint/typescript-eslint#6261]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.4&sourceType=module&code=DYUwLgBAbghgTgSxgI1BAvBARCgxliGAZwiLEQDsBzCAH2gHsEATAbgFgAoL2RFUAPwA6ANoAGALqsgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A

### Repro Code

```typescript
let variable = "abc" as string | void;

variable?.[0];
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No problems

### Actual Result

The third line reports an error - `Unnecessary optional chain on a non-nullish value.`

### Additional Info

Related to #5766. That PR did not extend to the optional chaining part of this rule.
This is caused by https://github.com/typescript-eslint/typescript-eslint/blob/ccd45d4a998946b7be1161f8c8216bc458e50b4e/packages/type-utils/src/predicates.ts#L28 not checking for `void`, only `null` and `undefined`.

**Comment by JoshuaKGoldberg** [comment]
👍 same as https://github.com/typescript-eslint/typescript-eslint/issues/5254#issuecomment-1166642504. This is a bug in the rule - though as per https://github.com/typescript-eslint/typescript-eslint/issues/5255 - you really should avoid using `| void`. See conversation in that issue for why we only reluctantly mark these as accepting PRs 😄
